File: LeetCode/Medium/Word Break/Java/wordBreak.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> words = new HashSet<>(wordDict);
        Queue<Integer> q = new LinkedList<>();
        boolean[] seen = new boolean[s.length()+1];
        q.add(0);
        while(!q.isEmpty()) {
            int start = q.remove();
            if (start == s.length()) return true;
            for (int end=start+1; end<=s.length(); end++) {
                if (seen[end]) continue;
                if (words.contains(s.substring(start, end))) {
                    q.add(end);
                    seen[end] = true;
                }
            }
        }
        return false;
    }
}

File: LeetCode/Medium/Word Break/C++/word_break.cc
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        set<string> words(wordDict.begin(), wordDict.end());
        bool seen[s.length()+1];
        for (int i=0; i<s.length()+1; i++) seen[i] = false;
        queue<int> q;
        q.push(0);
        while(!q.empty()) {
            int start = q.front();
            q.pop();
            if (start == s.length()) return true;
            for (int end=start+1; end<=s.length(); end++) {
                if (seen[end]) continue;
                string str = s.substr(start, end-start);
                if (words.find(str) != words.end()) {
                    q.push(end);
                    seen[end] = true;
                }
            }
        }
        return false;
    }
};
